File: Samples/WindowsML/Shared/cpp/ExecutionProviderManager.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE.md in the repo root for license information.
#include "ExecutionProviderManager.h"
#include "ArgumentParser.h"
#include <iostream>
#include <iomanip>
#include <unordered_map>
#include <winrt/Windows.Foundation.h>
#include <algorithm>
#include <winrt/base.h>

namespace {
// UTF-16 (wstring) -> UTF-8 using C++/WinRT helper
std::string WideToUtf8(const std::wstring& wide)
{
    return winrt::to_string(winrt::hstring{ wide });
}
}

namespace WindowsML
{
namespace Shared
{

    void ExecutionProviderManager::InitializeProviders(bool allowDownload)
    {
        std::cout << "Getting available providers..." << std::endl;
        auto catalog = winrt::Microsoft::Windows::AI::MachineLearning::ExecutionProviderCatalog::GetDefault();
        auto providers = catalog.FindAllProviders();
        for (const auto& provider : providers)
        {
            std::wcout << L"Provider: " << provider.Name().c_str() << std::endl;
            try
            {
                auto readyState = provider.ReadyState();
                std::wcout << L"  Ready state: " << static_cast<int>(readyState) << std::endl;

                // Only call EnsureReadyAsync if we allow downloads or if the provider is already ready
                if (allowDownload || readyState != winrt::Microsoft::Windows::AI::MachineLearning::ExecutionProviderReadyState::NotPresent)
                {
                    provider.EnsureReadyAsync().get();
                }

                provider.TryRegister();
            }
            catch (...)
            {
                // Continue if provider fails to initialize
            }
        }
    }

    bool ExecutionProviderManager::ConfigureSelectedExecutionProvider(Ort::SessionOptions& session_options,
                                                                      Ort::Env& env,
                                                                      const std::wstring& ep_name,
                                                                      const std::optional<std::wstring>& device_type)
    {
        // Discover devices
        std::vector<Ort::ConstEpDevice> ep_devices = env.GetEpDevices();

        // Accumulate devices by ep_name
        std::unordered_map<std::string, std::vector<Ort::ConstEpDevice>> ep_device_map;
        for (const auto& device : ep_devices)
        {
            std::string name(device.EpName());
            ep_device_map[name].push_back(device);
        }

        if (ep_name.empty())
        {
            std::wcout << L"ERROR: --ep_name is required when not using --ep_policy." << std::endl;
            PrintExecutionProviderInfo(ep_devices);
            return false;
        }

        const std::string ep_name_utf8 = WideToUtf8(ep_name);
        auto it = ep_device_map.find(ep_name_utf8);
        if (it == ep_device_map.end())
        {
            std::cout << "ERROR: Execution provider '" << ep_name_utf8 << "' not found among discovered devices." << std::endl;
            PrintExecutionProviderInfo(ep_devices);
            return false;
        }

        const auto& devices_for_ep = it->second;
        std::vector<Ort::ConstEpDevice> selected_devices;

        if (device_type.has_value())
        {
            std::cout << "Searching for device type '" << WideToUtf8(*device_type) << "' for provider '" << ep_name_utf8 << "'."
                      << std::endl;

            const std::string requested_type = WideToUtf8(*device_type);
            auto devices_itr = std::find_if(devices_for_ep.begin(), devices_for_ep.end(), [&](const Ort::ConstEpDevice& d) {
                return ArgumentParser::ToString(d.Device().Type()) == requested_type;
            });
            if (devices_itr == devices_for_ep.end())
            {
                std::cout << "ERROR: Device type '" << requested_type << "' not found for provider '" << ep_name_utf8 << "'." << std::endl;
                return false; // bail out
            }
            selected_devices.push_back(*devices_itr);
            std::cout << "Using device type '" << requested_type << "' for provider '" << ep_name_utf8 << "'." << std::endl;
        }
        else
        {
            // No device type specified: use all devices for the EP
            selected_devices = devices_for_ep;
            std::cout << "No device type specified; using the following device(s) for provider '" << ep_name_utf8 << "':" << std::endl;
            for (const auto& d : selected_devices)
            {
                std::cout << "  " << ArgumentParser::ToString(d.Device().Type()) << std::endl;
            }
        }

        // Finally, call append with the selected devices
        Ort::KeyValuePairs ep_options;
        try
        {
            session_options.AppendExecutionProvider_V2(env, selected_devices, ep_options);
            std::cout << "Successfully added " << ep_name_utf8 << " execution provider" << std::endl;
            return true;
        }
        catch (const Ort::Exception& ex)
        {
            std::cout << "ERROR: Failed to append execution provider '" << ep_name_utf8 << "': " << ex.what() << std::endl;
            return false;
        }
    }

    void ExecutionProviderManager::PrintExecutionProviderInfo(const std::vector<Ort::ConstEpDevice>& ep_devices)
    {
        std::unordered_map<std::string, std::vector<Ort::ConstEpDevice>> ep_device_map;
        for (const auto& device : ep_devices)
        {
            std::string name(device.EpName());
            ep_device_map[name].push_back(device);
        }

        std::wcout << L"Execution Provider Information:\n";
        std::wcout << L"-------------------------------------------------------------\n";
        std::wcout << std::left << std::setw(32) << L"Provider" << std::setw(16) << L"Vendor" << std::setw(12) << L"Device Type" << std::endl;
        std::wcout << L"-------------------------------------------------------------\n";
        for (auto& kv : ep_device_map)
        {
            const std::string& ep_name = kv.first;
            auto& devices = kv.second;
            for (const auto& device : devices)
            {
                std::cout << std::left << std::setw(32) << ep_name
                          << std::setw(16) << device.EpVendor() << std::setw(12)
                          << ArgumentParser::ToString(device.Device().Type()) << std::endl;
            }
        }
        std::wcout << L"-------------------------------------------------------------\n";
    }

    void ExecutionProviderManager::PrintExecutionProviderHelpTable()
    {
        try
        {
            Ort::Env env{};
            auto devices = env.GetEpDevices();
            if (!devices.empty())
            {
                PrintExecutionProviderInfo(devices);
            }
            else
            {
                std::wcout << L"(No execution provider devices discovered)\n";
            }
        }
        catch (...)
        {
            std::wcout << L"(Failed to query execution provider devices)\n";
        }
    }

} // namespace Shared
} // namespace WindowsML

File: Samples/WindowsML/Shared/cs/ExecutionProviderManager.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE.md in the repo root for license information.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.ML.OnnxRuntime;
using Microsoft.Windows.AI.MachineLearning;

namespace WindowsML.Shared
{
    /// <summary>
    /// Execution Provider configuration and discovery
    /// </summary>
    public static class ExecutionProviderManager
    {
        /// <summary>
        /// Initialize execution providers with download control
        /// </summary>
        public static async Task InitializeProvidersAsync(bool allowDownload = false)
        {
            Console.WriteLine("Getting available providers...");
            var catalog = ExecutionProviderCatalog.GetDefault();
            var providers = catalog.FindAllProviders();

            if (providers is null || providers.Length == 0)
            {
                Console.WriteLine("No execution providers found in catalog.");
                return;
            }

            foreach (var provider in providers)
            {
                Console.WriteLine($"Provider: {provider.Name}");
                try
                {
                    var readyState = provider.ReadyState;
                    Console.WriteLine($"  Ready state: {readyState}");

                    // Only call EnsureReadyAsync if we allow downloads or if the provider is already ready
                    if (allowDownload || readyState != ExecutionProviderReadyState.NotPresent)
                    {
                        await provider.EnsureReadyAsync();
                    }

                    provider.TryRegister();
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"  Failed to initialize provider {provider.Name}: {ex.Message}");
                    // Continue with other providers
                }
            }
        }
        /// <summary>
        /// Print execution provider device information
        /// </summary>
        public static void PrintDeviceInfo(Dictionary<string, List<OrtEpDevice>> epDeviceMap)
        {
            // Print table headers
            Console.WriteLine("Execution Provider Information:");
            Console.WriteLine("-------------------------------------------------------------");
            Console.WriteLine("{0,-32} {1,-16} {2,-12}", "Provider", "Vendor", "Device Type");
            Console.WriteLine("-------------------------------------------------------------");

            // Print each device with its provider name
            foreach (KeyValuePair<string, List<OrtEpDevice>> epGroup in epDeviceMap)
            {
                foreach (var device in epGroup.Value)
                {
                    Console.WriteLine("{0,-32} {1,-16} {2,-12}",
                        device.EpName,
                        device.EpVendor,
                        device.HardwareDevice.Type.ToString());
                }
            }

            Console.WriteLine("-------------------------------------------------------------");
        }

        /// <summary>
        /// Configure one explicitly selected execution provider (optional device type if provider exposes multiple).
        /// Returns true on success, false otherwise (prints guidance on failure).
        /// </summary>
        public static bool ConfigureSelectedExecutionProvider(SessionOptions sessionOptions,
                                                              OrtEnv environment,
                                                              string epName,
                                                              string? deviceType)
        {
            // Discover devices
            IReadOnlyList<OrtEpDevice> epDevices = environment.GetEpDevices();

            // Accumulate devices by EP name
            var epDeviceMap = new Dictionary<string, List<OrtEpDevice>>(StringComparer.OrdinalIgnoreCase);
            foreach (var d in epDevices)
            {
                if (!epDeviceMap.TryGetValue(d.EpName, out var list))
                {
                    list = new List<OrtEpDevice>();
                    epDeviceMap[d.EpName] = list;
                }
                list.Add(d);
            }

            if (string.IsNullOrWhiteSpace(epName))
            {
                Console.WriteLine("ERROR: --ep_name is required when not using --ep_policy.");
                PrintDeviceInfo(epDeviceMap);
                return false;
            }

            if (!epDeviceMap.TryGetValue(epName, out var devicesForEp))
            {
                Console.WriteLine($"ERROR: Execution provider '{epName}' not found among discovered devices.");
                PrintDeviceInfo(epDeviceMap);
                return false;
            }

            var selectedDevices = new List<OrtEpDevice>();

            if (!string.IsNullOrWhiteSpace(deviceType))
            {
                Console.WriteLine($"Searching for device type '{deviceType}' for provider '{epName}'.");
                var match = devicesForEp.FirstOrDefault(d => string.Equals(d.HardwareDevice.Type.ToString(), deviceType, StringComparison.OrdinalIgnoreCase));
                if (match == null)
                {
                    Console.WriteLine($"ERROR: Device type '{deviceType}' not found for provider '{epName}'.");
                    return false;
                }
                selectedDevices.Add(match);
                Console.WriteLine($"Using device type '{deviceType}' for provider '{epName}'.");
            }
            else
            {
                // No device type specified: use all devices for this EP
                selectedDevices.AddRange(devicesForEp);
                Console.WriteLine($"No device type specified; using the following device(s) for provider '{epName}':");
                foreach (var d in selectedDevices)
                {
                    Console.WriteLine($"  {d.HardwareDevice.Type}");
                }
            }

            try
            {
                var epOptions = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
                sessionOptions.AppendExecutionProvider(environment, selectedDevices, epOptions);
                Console.WriteLine($"Successfully added {epName} execution provider");
                return true;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"ERROR: Failed to append execution provider '{epName}': {ex.Message}");
                return false;
            }
        }

        public static async Task PrintExecutionProviderHelpTable()
        {
            try
            {
                // Call InitializeProvidersAsync first to ensure any providers from the catalog are registered and show up in the list
                await InitializeProvidersAsync(allowDownload: false);

                EnvironmentCreationOptions envOptions = new() {
                };

                var env = OrtEnv.CreateInstanceWithOptions(ref envOptions);
                var devices = env.GetEpDevices();
                if (devices.Count == 0)
                {
                    Console.WriteLine("(No execution provider devices discovered)");
                    return;
                }
                var epDeviceMap = new Dictionary<string, List<OrtEpDevice>>(StringComparer.OrdinalIgnoreCase);
                foreach (var d in devices)
                {
                    if (!epDeviceMap.TryGetValue(d.EpName, out var list))
                    {
                        list = new List<OrtEpDevice>();
                        epDeviceMap[d.EpName] = list;
                    }
                    list.Add(d);
                }
                PrintDeviceInfo(epDeviceMap);
            }
            catch
            {
                Console.WriteLine("(Execution provider enumeration unavailable in help context)");
            }
        }
    }
}
